### PR TITLE
[6.0][sending] Add support for Task.init and friends and TaskGroup and friends to take sending closures

### PR DESCRIPTION
**Pull request by gottesmm**
Explanation: This PR marks Task.init, Task.detached/et al and all the TaskGroup APIs as taking their operation as a sending non-Sendable parameter instead of as a @Sendable parameter. This allows for users to capture non-Sendable values into the closures used to call these APIs. E.x. we can write the following example which we couldn't before:

```
func example() {
  let x = NonSendableKlass()
  Task {
      useX(x)
   }
}
```

The rest of the changes in this PR were needed to work around other issues in the compiler that prevented us from doing the above.

Radars:

- rdar://130171939

Original PRs:

- https://github.com/swiftlang/swift/issues/74382

Risk: Low.
Testing: Added tests to the compiler.
Reviewer: N/A

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
Original: https://github.com/swiftlang/swift/pull/74239

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -2254,16 +2254,19 @@ ParamSpecifierRequest::evaluate(Evaluator &evaluator,
   if (auto transferring = dyn_cast<TransferringTypeRepr>(nestedRepr)) {
     // If we do not have an Ownership Repr, return implicit copyable consuming.
     auto *base = transferring->getBase();
-    if (!isa<OwnershipTypeRepr>(base)) {
+    if (!param->getInterfaceType()->isNoEscape() &&
```
Comment (xedin) [inline lib/Sema/TypeCheckDecl.cpp:2257]: Could you please double-check that this won't cause request dependency cycles with InterfaceTypeRequest?

**Review by xedin — APPROVED** [review]
Sema bits look reasonable to me.

**Review by hborla — APPROVED** [review]

File: lib/Sema/TypeCheckDecl.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/Sema/TypeCheckDecl.cpp:2257]: I checked. I didn't see anything obvious.

File: stdlib/public/Concurrency/Task.swift [hunk 2]
```diff
@@ -255,6 +255,8 @@ extension Task where Failure == Error {
     @_spi(MainActorUtilities)
     @MainActor
     @available(SwiftStdlib 5.9, *)
+    @available(*, deprecated, message: "Use Task.init with a main actor isolated closure instead")
+    @available(swift, obsoleted: 6.0, message: "Use Task.init with a main actor isolated closure instead")
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:259]: Thanks!

**Review by ktoso — COMMENTED** [review]
Looks good! Just double checked the APIs where we applied this.

Thanks Michael :)

**Pull request merged**